TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -212,40 +190,75 @@ - (TWTJSONSchemaNumberASTNode *)parseNumberSchema:(NSDictionary *)numberSchema
-
     node.maximumLength = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMaxLength schema:stringSchema];
     node.minimumLength = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMinLength schema:stringSchema];
-    // Does not check validity of regular expression, because definition is: "MUST be a string. This string SHOULD be a valid Regular Expression."
     node.pattern = [self parseStringForKey:TWTJSONSchemaKeywordPattern schema:stringSchema];
-
-    return node;
+    if (node.pattern) {
+        NSError *error = nil;
+        NSRegularExpression *regularExpression = [NSRegularExpression regularExpressionWithPattern:node.pattern options:0 error:&error];
+        if (error) {
+            [self warnWithFormat:@"Pattern %@ is not a valid regular expression, so it will not be used for validation. (All instances will pass.)", node.pattern];
+        } else {
+            node.regularExpression = regularExpression;
+        }
+    }
 }
 
 
-- (TWTJSONSchemaObjectASTNode *)parseObjectSchema:(NSDictionary *)objectSchema
+- (void)parseObjectSchema:(NSDictionary *)objectSchema intoNode:(TWTJSONSchemaObjectASTNode *)node
 {
-    TWTJSONSchemaObjectASTNode *node = [[TWTJSONSchemaObjectASTNode alloc] init];
-    [self parseCommonKeywordsFromSchema:objectSchema intoNode:node];
-
     node.maximumPropertyCount = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMaxProperties schema:objectSchema];
     node.minimumPropertyCount = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMinProperties schema:objectSchema];
-    NSArray *requiredNames = [self parseNonEmptyArrayOfUnqiueStringsForKey:TWTJSONSchemaKeywordRequired schema:objectSchema];
-    node.requiredPropertyNames = requiredNames ? [NSSet setWithArray:requiredNames] : nil;
+    NSArray *requiredKeys = [self parseNonEmptyArrayOfUnqiueStringsForKey:TWTJSONSchemaKeywordRequired schema:objectSchema];
+    node.requiredPropertyKeys = requiredKeys ? [NSSet setWithArray:requiredKeys] : nil;
     node.propertySchemas = [self parseDictionaryOfSchemasForKey:TWTJSONSchemaKeywordProperties schema:objectSchema keyValuePairNodeClass:[TWTJSONSchemaNamedPropertyASTNode class]];
     node.patternPropertySchemas = [self parseDictionaryOfSchemasForKey:TWTJSONSchemaKeywordPatternProperties schema:objectSchema keyValuePairNodeClass:[TWTJSONSchemaPatternPropertyASTNode class]];
     node.additionalPropertiesNode = [self parseAdditionalItemsOrPropertiesForKey:TWTJSONSchemaKeywordAdditionalProperties schema:objectSchema];
     node.propertyDependencies = [self parseDependenciesWithSchema:objectSchema];
+}
 
-    return node;
+
+- (void)parseAmbiguousSchema:(NSDictionary *)ambigousSchema intoNode:(TWTJSONSchemaAmbiguousASTNode *)node withTypes:(NSArray *)types
+{
+    node.validTypes = [NSSet setWithArray:types];
+
+    NSMutableArray *subNodes = [[NSMutableArray alloc] init];
+    for (NSString *type in types) {
+        id subNode = nil;
```

Comment:
Since subNode will always be a `TWTJSONSchemaASTNode`, you should use that type instead of `id`. Basically, always use the most specific type you can so that the compiler can help you from making mistakes.